### PR TITLE
Don't overwrite custom etag

### PR DESCRIPTION
**Pull request by cristovao-trevisan**
Kit should not overwrite custom etag value.

Sometimes it's useful to provide a custom etag value which should not be overwritten by sveltekit.
For example, in my particular use case, when proxying an image from S3 it's possible to retrieve the ETAG with the metadata, without downloading it.
This is also useful when the etag is based on api requests, making it possible to skip rendering the page.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpx changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

**Comment by Rich-Harris** [comment]
Thank you! Can you add a test and a changeset please? You could add a test to this section: https://github.com/sveltejs/kit/blob/9e482ac43b2d3f903f497ab5c9cc92691aec9906/packages/kit/test/apps/basics/test/test.js#L690-L735

**Comment by Conduitry** [comment]
Should we also, then, trust an `etag` from the user's `response` object when we're checking whether we should return a 304?

**Comment by cristovao-trevisan** [comment]
Yes @Rich-Harris , I can add a test and the changeset, will do soon

@Conduitry , I don't think there's a stateless (or serveless) way to check the user provided etag, I assume that if the etag is being set manually the user know what he's doing and will check the if-none-matches equality himself

File: .changeset/clever-readers-turn.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+'@sveltejs/kit': patch
+---
+
+Don't overwrite ETAG header.
+
+Now user can provide a custom ETAG header, which should be manually checked against the 'if-none-match' header.
```
Comment (Rich-Harris) [inline .changeset/clever-readers-turn.md:7]:
```suggestion
Preserve explicit ETag header
```

File: packages/kit/src/runtime/server/index.js [hunk 2]
```diff
@@ -152,7 +152,9 @@ export async function respond(incoming, options, state = {}) {
 									};
 								}
 
-								response.headers['etag'] = etag;
+								if (!response.headers['etag']) {
+									response.headers['etag'] = etag;
+								}
```
Comment (Rich-Harris) [inline packages/kit/src/runtime/server/index.js:157]: Shouldn't we be bailing out on line 121? If there's already an ETag then surely we should never get to this point?

```suggestion
								response.headers['etag'] = etag;
```

Feels like we need to make this change instead:

```diff
-if (response.status === 200) {
+if (response.status === 200 && !response.headers['etag']) {

**Review by Rich-Harris — CHANGES_REQUESTED** [review]
Thanks — response inline

**Review by Rich-Harris — APPROVED** [review]

File: packages/kit/src/runtime/server/index.js (hunk 2 above)
Comment (Rich-Harris) [reply packages/kit/src/runtime/server/index.js]: made this change

**Comment by cristovao-trevisan** [comment]
Was going to apply your suggestion right now, sorry to be late but thanks for all the work!!

**Pull request merged**